src/Task/ProtectedTaskManager.cpp

```diff
@@ -84,29 +85,36 @@ ProtectedTaskManager::IncrementActiveTaskPointArm(int offset)
 {
   ExclusiveLease lease(*this);
   TaskAdvance &advance = lease->SetTaskAdvance();
+  IntermediateTaskPoint *nextwp = nullptr;
 
   switch (advance.GetState()) {
   case TaskAdvance::MANUAL:
   case TaskAdvance::AUTO:
     lease->IncrementActiveTaskPoint(offset);
+    nextwp = (IntermediateTaskPoint *)lease->GetActiveTaskPoint();
     break;
   case TaskAdvance::START_DISARMED:
   case TaskAdvance::TURN_DISARMED:
     if (offset>0) {
       advance.SetArmed(true);
     } else {
       lease->IncrementActiveTaskPoint(offset);
+      nextwp = (IntermediateTaskPoint *)lease->GetActiveTaskPoint();
     }
     break;
   case TaskAdvance::START_ARMED:
   case TaskAdvance::TURN_ARMED:
     if (offset>0) {
       lease->IncrementActiveTaskPoint(offset);
+      nextwp = (IntermediateTaskPoint *)lease->GetActiveTaskPoint();
```

Comment:
How do you know that these casts are legal?

Reply:
Diving through the src tree, lot of "grep" and paper.
Not a simple task and I hope I didn't make a mistake. I wish I had a static code analyzer, for instance to follow inheritance. But I don't. All I have (and know) is gdb.

Reply:
So all your code diving made you confident that `GetActiveTaskPoint()` will only ever return `IntermediateTaskPoint` instances? And never a `FinishPoint`?

Reply:
Good point. But it never moves past the FinishPoint.

`OrderedTaskPoint` might be more appropriate.
I need to think about it.

Reply:
Here is the result of my thinking, @MaxKellermann :

Prerequisite:

1. `class ScoredTaskPoint` has the methods I need: `Reset()` and `HasEntered()` .

2. `class OrderedTaskPoint` inherits from `ScoredTaskPoint` and `TaskWaypoint` , and from others.
```
class OrderedTaskPoint
  :public TaskLeg,
   public TaskWaypoint,
   public ScoredTaskPoint,
   public ObservationZoneClient
{ .....
```

3. In `IncrementActiveTaskPointArm()` I have a `TaskWaypoint*` which points to the object I want to work on.

**Question**: Can I simply type cast a `TaskWaypoint*` to `OrderedTaskPoint*` if I want access to the methods described in #1 ?
When I think about what a compiler can do then I guess the answer is YES.

Reply:
You can always cast everything to everything. The compiler will "allow" everything. The compiler doesn't know if it's a good idea. Casts are however only a good idea if you really know the object is really an instance of the target type. This you need to make sure before you cast.
If you're dealing with an `OrderedTask`, all points will always be `OrderedTaskPoint`, which is the base class for all other point types in ordered tasks. Therefore, after you have ensured that this is an `OrderedTask`, casting to `OrderedTaskPoint` is legal.